src/execution_plan/ops/shared/update_functions.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2018-2021 Redis Labs Ltd. and Contributors
+ *
+ * This file is available under the Redis Labs Source Available License Agreement
+ */
+
+#include "update_functions.h"
+#include "../../../errors.h"
+
+/* Set a property on a graph entity. For non-NULL values, the property
+ * will be added or updated if it is already present.
+ * For NULL values, the property will be deleted if present
+ * and nothing will be done otherwise.
+ * Returns 1 if a property was set or deleted.  */
+static int _UpdateEntity(PendingUpdateCtx *update) {
+	int           res        =  0;
+	GraphEntity   *ge        =  update->ge;
+	Attribute_ID  attr_id    =  update->attr_id;
+	SIValue       new_value  =  update->new_value;
+
+	// If this entity has been deleted, perform no updates and return early.
+	if(GraphEntity_IsDeleted(ge)) goto cleanup;
+
+	// Handle the case in which we are deleting all properties.
+	if(attr_id == ATTRIBUTE_ALL) return GraphEntity_ClearProperties(ge);
+
+	// Try to get current property value.
+	SIValue *old_value = GraphEntity_GetProperty(ge, attr_id);
+
+	if(old_value == PROPERTY_NOTFOUND) {
+		// Adding a new property; do nothing if its value is NULL.
+		if(SI_TYPE(new_value) != T_NULL) {
+			res = GraphEntity_AddProperty(ge, attr_id, new_value);
+		}
+	} else {
+		// Update property.
+		res = GraphEntity_SetProperty(ge, attr_id, new_value);
+	}
+
+cleanup:
+	SIValue_Free(new_value);
+	return res;
+}
+
+// Commits delayed updates.
+void CommitUpdates(GraphContext *gc, ResultSetStatistics *stats, PendingUpdateCtx *updates) {
+	uint update_count = array_len(updates);
+	uint properties_set = 0;
+	for(uint i = 0; i < update_count; i++) {
+		PendingUpdateCtx *update = updates + i;
+		// Update the property on the graph entity.
+		properties_set += _UpdateEntity(update);
+
+		// Update index for node entities if indexed fields have been modified.
+		if(update->update_index) {
+			Schema *s = GraphContext_GetSchemaByID(gc, update->label_id, SCHEMA_NODE);
+			// Introduce updated entity to index.
+			Schema_AddNodeToIndices(s, (Node *)update->ge);
+		}
```

Comment:
```
CREATE INDEX ON :Person(age)
CREATE INDEX ON :Person(name)
MATCH (p:Person) SET p.age = 33, p.name = 'Alex'
```

Update query updates both indexed fields, but we should call `Schema_AddNodeToIndices` only once to re-index the entity, otherwise there will be a performance hit.

this might be a simple fix, checking if we need to update the index within the loop, but performing the test and calling reindexing outside of the loop:

```
bool reindex = false;
for(uint i = 0; i < update_count; i++) {
    ...
    reindex |= update->update_index;
}

if(reindex) ...
```